src/main/java/com/rultor/agents/docker/StartsDockerDaemon.java

```diff
@@ -0,0 +1,148 @@
+    /**
+     * Profile used.
+     */
+    private final transient Profile profile;
+
+    /**
+     * Docker client.
+     */
+    private final transient DockerClient client;
+
+    /**
+     * Docker Containers created.
+     */
+    private final transient Collection<CreateContainerResponse> containers;
+
+    /**
+     * Ctor.
+     * @param prof Current Profile
+     */
+    public StartsDockerDaemon(final Profile prof) {
+        this.profile = prof;
+        this.client = DockerClientBuilder.getInstance().build();
+        this.containers = Collections.newSetFromMap(
+            // @checkstyle MagicNumber (1 line)
+            new ConcurrentHashMap<CreateContainerResponse, Boolean>(1, 0.9f, 1)
+        );
+    }
+
+    /**
+     * Sets up a PfShell in which Rultor can function.
+     * @return PfShell loaded with credentials for new Rultor runner container
+     * @throws IOException on failure
+     */
+    public PfShell shell() throws IOException {
+        final ExposedPort ssh = ExposedPort.tcp(22);
+        final Ports ports = new Ports();
+        ports.bind(ssh, Ports.Binding(null));
+        final CreateContainerResponse container = this.client
+            .createContainerCmd(this.build()).withExposedPorts(
+                ssh
+            ).withPortBindings(ports).exec();
+        this.containers.add(container);
+        this.client.startContainerCmd(container.getId()).exec();
+        return new PfShell(
+            this.profile,
+            this.client.infoCmd().exec().getName(),
+            this.client.inspectContainerCmd(container.getId())
+                .exec().getNetworkSettings().getPorts().getBindings().values()
+                .iterator().next()[0].getHostPort(),
+            "root",
+            this.key(container)
+        );
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (final CreateContainerResponse container : this.containers) {
+            this.client.killContainerCmd(container.getId()).exec();
+            this.client.removeContainerCmd(container.getId()).exec();
+        }
+        this.client.close();
+    }
+
+    /**
+     * Retrieves SSH private key needed to connect to a given container.
+     * @param container Container from which to get the SSH key
+     * @return SSH private key
+     * @throws IOException on failure
+     */
+    private String key(final CreateContainerResponse container)
+        throws IOException {
+        final StringWriter writer = new StringWriter();
+        IOUtil.copy(
+            this.client.copyFileFromContainerCmd(
+                container.getId(), "/root/.ssh/id_rsa"
+            ).exec(),
+            writer
+        );
+        final String key = writer.toString();
+        return key.substring(key.indexOf('-'), key.lastIndexOf('-') + 1);
+    }
+
+    /**
+     * Builds a fresh Rultor runner base image on the Docker daemon.
+     * @return Image ID of the Rultor runner base image
+     */
+    private String build() {
```

Comment:
@mkordas <10s on he second run and beyond. Only the first time this runs on a new Docker daemon takes a few min. After that it's almost instant with the build cache.